### PR TITLE
feat(LERN-93): implemented user list for getGrades endpoint (admin)

### PR DESCRIPTION
**Pull request by A1eeee**
Fix: fixed getTeachers query for admins

### CODE REVIEW

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeService.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import ch.learntrack.backend.utils.sanitizeInputString
+import java.time.LocalDateTime
+import java.util.UUID
+
+public class AdminGradeService(
+    private val gradeService: GradeService,
+    private val userService: UserService,
+    private val gradeDao: GradeDao,
+) {
+    public fun getGrades(schoolId: UUID): GradeDetailsDto = GradeDetailsDto(
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeService.kt:20]:
```suggestion
    public fun getAllGradesForSchool(schoolId: UUID): GradeDetailsDto = GradeDetailsDto(
```

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeService.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            id = UUID.randomUUID(),
+            name = createGradeDto.name.trim().sanitizeInputString(),
+            schoolId = createGradeDto.schoolId,
+            created = LocalDateTime.now(),
+            updated = LocalDateTime.now(),
+        )
+
+        gradeDao.insert(grade)
+
+        return gradeService.mapToDto(grade)
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeService.kt:40]: why are you returning the grade here? Is there any specific reason?

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeBeans.kt [hunk 3]
```diff
@@ -5,5 +5,5 @@ import org.springframework.context.support.beans
 
 public val gradeBeans: BeanDefinitionDsl = beans {
     bean<GradeResource>()
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeBeans.kt:7]: I would probably rename the Resource to `AdminGradeResource` as well.

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeDetailsDto.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package ch.learntrack.backend.admin.grade
+
+import ch.learntrack.backend.grade.GradeDto
+import ch.learntrack.backend.user.UserDto
+
+public data class GradeDetailsDto(
+    val grades: List<GradeDto>,
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeDetailsDto.kt:7]: That shouldn't be a list.

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeDetailsDto.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+    val users: List<UserDto>,
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeDetailsDto.kt:8]:
```suggestion
    val students: List<UserDto>,
```

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeResource.kt [hunk 6]
```diff
@@ -14,21 +15,21 @@ import java.util.UUID
...
-    private val gradeService: GradeService,
+    private val adminGradeService: AdminGradeService,
 ) {
     @PreAuthorize(
         "@UserAccessAuthorizer.hasUserAccessToSchool(#root.authentication.principal, #schoolId)",
     )
     @GetMapping
-    public fun getGrades(@RequestParam schoolId: UUID): List<GradeDto> =
-        gradeService.getGrades(schoolId)
+    public fun getGrades(@RequestParam schoolId: UUID): GradeDetailsDto =
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeResource.kt:24]:
```suggestion
    public fun getGrades(@RequestParam schoolId: UUID): List<GradeDetailsDto> =
```

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeResource.kt [hunk 7]
```diff
@@ -14,21 +15,21 @@ import java.util.UUID
...
+        adminGradeService.getGrades(schoolId)
 
     @PreAuthorize(
         "@UserAccessAuthorizer.hasUserAccessToSchool(#root.authentication.principal, #schoolId)",
     )
     @PostMapping("/createGrade")
-    public fun createGrade(@RequestParam schoolId: UUID, @RequestBody createGradeDto: CreateGradeDto): Grade =
-        gradeService.createGrade(
+    public fun createGrade(@RequestParam schoolId: UUID, @RequestBody createGradeDto: CreateGradeDto): GradeDto =
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/GradeResource.kt:31]: I don't think we need/want to return the grade again. the frontend has to refetch anyway after creating a grade to update the grade list.

File: backend/src/main/kotlin/ch/learntrack/backend/admin/teacher/TeacherBeans.kt [hunk 8]
```diff
@@ -5,5 +5,5 @@ import org.springframework.context.support.beans
 
 public val teacherBeans: BeanDefinitionDsl = beans {
     bean<TeacherResource>()
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/teacher/TeacherBeans.kt:7]: Maybe also rename this to `AdminTeacherResource`?

File: backend/src/main/kotlin/ch/learntrack/backend/grade/GradeDto.kt [hunk 9]
```diff
@@ -7,3 +7,8 @@ public data class GradeDto(
     val name: String,
     val schoolId: UUID,
 )
+
+public data class CreateGradeDto(
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/grade/GradeDto.kt:11]: Maybe move this to the admin grade package...  as we won't need it again for something else?

File: backend/src/main/kotlin/ch/learntrack/backend/grade/GradeService.kt [hunk 10]
```diff
@@ -17,4 +18,14 @@ public class GradeService(private val gradeDao: GradeDao) : EntityService<GradeD
     public fun getGradesForUserByUserId(userId: UUID): List<GradeDto> = gradeDao.fetchGradesForUserByUserId(userId).map(
         ::mapToDto,
     )
+
+    public fun getGradesBySchoolId(schoolId: UUID): List<GradeDto> = gradeDao.fetchBySchoolId(schoolId).map(
+        ::mapToDto)
+
+    public fun getGradeByNameAndSchoolId(createGradeDto: CreateGradeDto): Grade? = gradeDao.fetchByNameAndSchoolId(
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/grade/GradeService.kt:25]: specifically give the 2 parameters `name` and `SchoolId`. in anyway it shouldn't be named `create`

File: backend/src/main/kotlin/ch/learntrack/backend/grade/GradeService.kt [hunk 11]
```diff
@@ -17,4 +18,14 @@ public class GradeService(private val gradeDao: GradeDao) : EntityService<GradeD
...
+        createGradeDto,
+    )
+
+    public fun isGradeNameExistingInThisSchool(createGradeDto: CreateGradeDto): Boolean =
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/grade/GradeService.kt:29]: same here give the parameters explicitly.

**Review by deepspace1000 — CHANGES_REQUESTED** [review]

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 0  | ❗       |

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeResource.kt [hunk 12]
```diff
@@ -13,22 +12,23 @@ import java.util.UUID
...
         "@UserAccessAuthorizer.hasUserAccessToSchool(#root.authentication.principal, #schoolId)",
     )
     @PostMapping("/createGrade")
-    public fun createGrade(@RequestParam schoolId: UUID, @RequestBody createGradeDto: CreateGradeDto): Grade =
-        gradeService.createGrade(
+    public fun createGrade(@RequestParam schoolId: UUID, @RequestBody createGradeDto: CreateGradeDto) {
+        adminGradeService.createGrade(
             createGradeDto,
         )
+    }
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeResource.kt:33]:
```suggestion
    public fun createGrade(@RequestParam schoolId: UUID, @RequestBody createGradeDto: CreateGradeDto) = adminGradeService.createGrade(
            createGradeDto,
        )
    
```

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeService.kt [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        gradeService.getGradesBySchoolId(schoolId).map { grade ->
+            GradeDetailsDto(
+                grades = grade,
+                students = userService.getUsersByRoleAndGradeId(UserRole.STUDENT, grade.id),
+                teachers = userService.getUsersByRoleAndGradeId(UserRole.TEACHER, grade.id),
+            )
+        }
+
+    public fun createGrade(createGradeDto: CreateGradeDto) {
+        if (gradeService.isGradeNameExistingInThisSchool(createGradeDto.schoolId, createGradeDto.name)) {
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeService.kt:28]:
```suggestion
        if (gradeService.isGradeExistingInSchool(createGradeDto.name, createGradeDto.schoolId)) {
```

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeService.kt [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+            throw LearnTrackConflictException("Grade with name ${createGradeDto.name} already exists in this school")
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeService.kt:29]:
```suggestion
            throw LearnTrackConflictException("Grade with name ${createGradeDto.name} already exists in school ${createGradeDto.shoolId}")
```

File: backend/src/main/kotlin/ch/learntrack/backend/grade/GradeService.kt [hunk 15]
```diff
@@ -17,4 +18,15 @@ public class GradeService(private val gradeDao: GradeDao) : EntityService<GradeD
...
+
+    public fun getGradesBySchoolId(schoolId: UUID): List<GradeDto> = gradeDao.fetchBySchoolId(schoolId).map(
+        ::mapToDto)
+
+    public fun getGradeByNameAndSchoolId(schoolId: UUID, name: String): Grade? = gradeDao.fetchByNameAndSchoolId(
+        schoolId,
+        name,
+    )
+
+    public fun isGradeNameExistingInThisSchool(schoolId: UUID, name: String): Boolean =
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/grade/GradeService.kt:30]:
```suggestion
    public fun isGradeExistingInSchool(name: String, schoolId: UUID): Boolean =
```

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeResource.kt [hunk 16]
```diff
@@ -13,22 +12,23 @@ import java.util.UUID
...
+public class AdminGradeResource(
+    private val adminGradeService: AdminGradeService,
 ) {
     @PreAuthorize(
         "@UserAccessAuthorizer.hasUserAccessToSchool(#root.authentication.principal, #schoolId)",
     )
     @GetMapping
-    public fun getGrades(@RequestParam schoolId: UUID): List<GradeDto> =
-        gradeService.getGrades(schoolId)
+    public fun getAllGradesForSchool(@RequestParam schoolId: UUID): List<GradeDetailsDto> =
```
Comment (deepspace1000) [inline backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeResource.kt:22]: maybe `getAllGradesWithDetailsBySchoolId`? 
Definitely use `BySchoolId`.
Maybe use `GradesWithDetails`.

File: backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeResource.kt (hunk 12 above)
Comment (A1eeee) [reply backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeResource.kt:33]: The problem is if we use it that way I need to specify a :Unit and sadly the diktatCheck doesn't like that he says that is unnecessary and if I don't use that I get an error from the Kotlin side that we need to use :Unity.

That's why we didn't use your suggestion if I remember correctly
 backend/src/main/kotlin/ch/learntrack/backend/backoffice/user/UserResource.kt

Comment (deepspace1000) [reply backend/src/main/kotlin/ch/learntrack/backend/admin/grade/AdminGradeResource.kt:33]: You can try to remove that from the diktat config in that case.

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 0  | ❗       |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 0  | ❗       |

File: backend/detekt.yml
Comment (deepspace1000) [inline backend/detekt.yml:1]: 🙃

**Review by deepspace1000 — COMMENTED** [review]
Are there any test yet?

**Comment by A1eeee** [comment]
> Are there any test yet?

I mean that task is a fix/enhancement task for the other two tasks that I've done before that one.

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/login-page*: 


| Before | After |
| --- | --- |
|  |  |

File: backend/src/test/kotlin/ch/learntrack/backend/admin/grade/GradeIntegrationTest.kt
Comment (deepspace1000) [inline backend/src/test/kotlin/ch/learntrack/backend/admin/grade/GradeIntegrationTest.kt:1]: add some more test data like second grade in a different school, which shouldn't be returned. 
add a teacher to 2 grades in 2 schools and check if he is returned in both. ect.

**Review by deepspace1000 — COMMENTED** [review]
FYI you can enhance your commit messages a bit e.g you last commit could be: `added tests for admin grade endpoints` or something like that. 
The message `implemented requested changes` can be used when there is nothing really to describe what you did but should not be used if possible.

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 1  | ❗       |
    


*❗ custom/login-page*: 


| Before | After |
| --- | --- |
|  |  |

**Review by deepspace1000 — APPROVED** [review]
LGTM

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 0  | ❗       |

**Pull request merged**